tmt/utils.py

```diff
@@ -374,20 +474,27 @@ def _run(
 
         # Handle the exit code, return output
         if process.returncode != 0:
-            if isinstance(command, (list, tuple)):
-                command = ' '.join(command)
             raise RunError(
                 message=f"Command returned '{process.returncode}'.",
                 command=command,
                 returncode=process.returncode,
                 stdout=stdout_thread.get_output(),
                 stderr=stderr_thread.get_output())
-        return stdout_thread.get_output() if join else (
+        return (stdout_thread.get_output(), None) if join else (
             stdout_thread.get_output(), stderr_thread.get_output())
 
-    def run(
-            self, command, message=None, cwd=None, dry=False, shell=False,
-            env=None, interactive=False, join=False, log=None, timeout=None):
+    def run(self,
+            command: List[str],
```

Comment:
If `shell=True` is used the method expects `command` to be a string, not a list.

Reply:
Would it be possible to keep this parameter `List[str]`, and use `shlex.join(command)` inside `run()` when `shell=True`?

According to documentation, it is _recommended_ for `args` to be a string with `shell=True`, and I find this duality puzzling. It's not a bug to use a list of strings with `shell=True`, it's just interpreted differently (`args[0]` a command to execute, everything else options for the shell itself) - I'd prefer seeing some kind of `shell_args` option rather than giving the main parameter different meaning. I guess I brought this view into this patch, "command shall be a list of strings" makes things easier to reason about - but to make `subprocess` happy, we can add `shlex.join()` and pass a joined string to `subprocess.run()`.

Is there a place where `run()` is called with `shell=True`? I've found one, https://github.com/teemtee/tmt/blob/main/tmt/steps/provision/local.py#L72, but I can't tell what's the type of `command` there. Would something like this be acceptable?

```diff
diff --git a/tmt/utils.py b/tmt/utils.py
index 56284b4..e521419 100644
--- a/tmt/utils.py
+++ b/tmt/utils.py
@@ -432,11 +432,21 @@ class Common(object):
             environment = None
         self.debug('environment', pprint.pformat(environment), level=4)
 
+        # subprocess.run() recommends using a string instead of list of strings when shell=True.
+        final_command: Union[str, List[str]] = command
+
+        if isinstance(command, str):
+            # do nothing - accept the string, but emit a warning because a command shall be a list of strings.
+            self.warn(f'command given as a string: {command}')
+
+        elif shell:
+            final_command = shlex.join(command)
+
         # Run the command in interactive mode if requested
         if interactive:
             try:
                 subprocess.run(
-                    command, cwd=cwd, shell=shell, env=environment, check=True)
+                    final_command, cwd=cwd, shell=shell, env=environment, check=True)
             except subprocess.CalledProcessError as error:
                 # Interactive mode can return non-zero if the last command
                 # failed, ignore errors here
@@ -447,7 +457,7 @@ class Common(object):
         # Create the process
         try:
             process = subprocess.Popen(
-                command, cwd=cwd, shell=shell, env=environment,
+                final_command, cwd=cwd, shell=shell, env=environment,
                 stdin=subprocess.DEVNULL, stdout=subprocess.PIPE,
                 stderr=subprocess.STDOUT if join else subprocess.PIPE)
         except FileNotFoundError as error:
```

Reply:
After a debate, resolved with 8b816ab.